### DISCUSSION TITLE
A small finding in TESPy Network Class and improvement suggestions

### DISCUSSION THREAD

**Discussion by GoldenPotatis** [oemof/tespy#359] [Q&A] [answered]
Hi!

I'm a new user of TESPy and still learning. I found this small issue and would like to share with the community.
I'm using Python3.8.9 and TESPy0.6.0

In the Network Class, it is possible to view the network connections and components by calling the `network.conns` attributes and `network.comps` attributes. The `network.conns` works perfectly fine. However, the user may find that `network.comps` does not work. It returns `AttributeError: 'Network' object has no attribute 'comps'`.

The example I'm using is a simple condenser cooled by air, same as in the [Condenser example code in tespy.readthedocs](https://tespy.readthedocs.io/en/main/api/tespy.components.html#module-tespy.components.heat_exchangers.condenser):

```python
from tespy.networks import Network
from tespy.components import Source, Sink, Condenser
from tespy.connections import Connection

# create a network and components
nw = Network(fluids=['water', 'air'], T_unit='C', p_unit='bar', h_unit='kJ / kg')
air_in = Source(label='air inlet')
air_out = Sink('air outlet')
waste_steam = Source('waste steam')
water = Sink('condensate water')
condenser = Condenser('condenser')

# create connections
air_con = Connection(air_in, 'out1', condenser, 'in2', label='air to condenser')
con_air = Connection(condenser, 'out2', air_out, 'in1', label='condenser to air')
ws_con = Connection(waste_steam, 'out1', condenser, 'in1', label='ws to con')
con_water = Connection(condenser, 'out1', water, 'in1')

nw.add_conns(air_con, con_air, ws_con, con_water)

```
When calling the `nw.conns` attributes, it will return a Pandas DataFrame. But calling `nw.comps` will result in `AttributeError: 'Network' object has no attribute 'comps'`.

![image](https://user-images.githubusercontent.com/30263046/191722120-8add9f70-7cc3-42a1-8498-dc3b3a799390.png)

Correspondingly, the `Network.get_comp()` doesn't work neither.

After looking into the source code of Network, it is found that the `network.conns` is initiated at the beginning under `def __init__(self, fluids, memorise_fluid_properties=True, **kwargs)` and `def set_defaults(self)`. But `network.comps` is not initiated. The first time `network.comps` is initiated is actually under `def check_network(self)`, which is:

```python
comps = pd.unique(self.conns[['source', 'target']].values.ravel())
# build the dataframe for components
```

After running `nw.check_network()`, the `nw.comps` then works perfectly fine, which gives a dataframe of the components.

![image](https://user-images.githubusercontent.com/30263046/191725300-efe870af-7092-41bd-a7fe-d511626e09e6.png)

It is therefore suggested to initiate the `.comps` at the beginning of the code to avoid such error.

Another attached question is under the example code of Condenser in readthedocs, why the air inlet is Sink and air outlet is Source?

```python
amb_in = Sink('ambient air inlet')
amb_out = Source('air outlet')
```

Based on the definition of Source and Sink, should it be like air inlet is Source and air outlet is Sink?

**Comment by fwitte** [comment, marked as answer]
This had been resolved by a previous PR #362
